components/net/http_cache.rs

```diff
@@ -0,0 +1,660 @@
+                constructed_response.status = cached_resource.status.clone();
+                constructed_response.https_state = cached_resource.https_state.clone();
+                constructed_response.referrer = request.referrer.to_url().cloned();
+                constructed_response.referrer_policy = request.referrer_policy.clone();
+                constructed_response.raw_status = cached_resource.raw_status.clone();
+                constructed_response.url_list = cached_resource.url_list.clone();
+                // done_chan will have been set to Some by http_network_fetch,
+                // set it back to None since the response returned here replaces the 304 one from the network.
+                *done_chan = None;
+                cached_resource.expires = get_response_expiry(&constructed_response);
+                return Some(constructed_response);
+            }
+        }
+        None
+    }
+
+    fn invalidate_for_url(&mut self, url: &ServoUrl) {
+        let entry_key = CacheKey::from_servo_url(url);
+        if let Some(cached_resources) = self.entries.get_mut(&entry_key) {
+            for cached_resource in cached_resources.iter_mut() {
+                cached_resource.expires = Duration::seconds(0i64);
+            }
+        }
+    }
+
+    /// Invalidation.
+    /// <https://tools.ietf.org/html/rfc7234#section-4.4>
+    pub fn invalidate(&mut self, request: &Request, response: &Response) {
+        if let Some(&header::Location(ref location)) = response.headers.get::<header::Location>() {
+            if let Ok(url) = request.current_url().join(location) {
+                self.invalidate_for_url(&url);
+            }
+        }
+        // TODO: update hyper to use typed getter.
+        if let Some(url_data) = response.headers.get_raw("Content-Location") {
+            if let Ok(content_location) = str::from_utf8(&url_data[0]) {
+                if let Ok(url) = request.current_url().join(content_location) {
+                    self.invalidate_for_url(&url);
+                }
+            }
+        }
+        self.invalidate_for_url(&request.url());
+    }
+
+    /// Storing Responses in Caches.
+    /// <https://tools.ietf.org/html/rfc7234#section-3>
+    pub fn store(&mut self, request: &Request, response: &Response) {
+        if let Some(status) = response.status {
+            // Not caching redirects, for simplicity, not per the spec.
+            if is_redirect_status(status) {
+                return
+            }
+        }
+        if request.method != Method::Get {
+            // Only Get requests are cached.
+            return
+        }
+        if let Some(data) = request.headers.get_raw("cache-control") {
+            let directives_string = String::from_utf8_lossy(&data[0]);
+            for directive in directives_string.split(",") {
+                let mut directive_info = directive.split("=");
+                if directive_info.next() == Some("no-cache") {
```

Comment:
I just realized what I am doing here isn't correct, because although the 'no-cache' directive is set by fetch for a Reload or Nostore cache mode(and so this makes the relevant tests pass), the directive itself is unrelated to cache storage, so instead I am looking at the cachemode logic in the fetch code, I think it needs to be slightly adapted so that the cache is used or not based on the cachemode as described in https://fetch.spec.whatwg.org/#concept-request-cache-mode